test/functional/test_framework/p2p.py

```diff
@@ -163,6 +163,7 @@ def __init__(self):
         # The underlying transport of the connection.
         # Should only call methods on this from the NetworkThread, c.f. call_soon_threadsafe
         self._transport = None
+        self._send_lock = threading.Lock()
```

Comment:
One thing to consider is that now we have two locks for p2p, we need to worry about deadlocks.
There is an implied lock order, `on_message` locks `p2p_lock`, and can call messages (e.g. `on_inv`) that will call `send_message` and lock `_send_lock`. So we must make sure never to take `p2p_lock` after `_send_lock`. I don't think that this is currently possible, but maybe we should add a comment about this.

Reply:
good find! i've added a comment.

Reply:
@mzumsande, @sr-gi, I played with @sr-gi's idea on [this branch](https://github.com/stratospher/bitcoin/commit/276b95d766cd11d65d563805d52beace8dd32744) and wondering about 2 things:

1. can deadlocks happen? i noticed places in the test where `_send_lock` was acquired and then `p2p_lock` (and deadlocks didn't happen). am i missing something?
	1. MainThread acquires `_send_lock`
	2. NetworkThread acquires `p2p_lock`
	3. NetworkThread would wait for MainThread to release `_send_lock`
	4. MainThread would release `_send_lock` at some point of time right? (since it doesn't need to wait for `p2p_lock`)
2. do we need a separate `_send_lock` for every P2PConnection? benchmarks look slightly faster with a global `_send_lock` (like `p2p_lock`).

> Given this will affect several parts of the tests suite it may be worth doing it as a followup

+1, but supposing deadlocks are possible - in the follow up:
1. `_send_lock` acquired because of `send_message()` and then `p2p_lock` acquired because of `wait_until()` is harmless for us
2. `_send_lock` acquired because of `send_message()` and then `p2p_lock` acquired because of `on_message()` is the situation we need to be concerned about

Reply:
> do we need a separate `_send_lock` for every P2PConnection? benchmarks look slightly faster with a global `_send_lock` (like `p2p_lock`).

Umm, that sounds a bit strange to me. My intuition would be that having a global lock should be slower, especially if multiple connections are run at the same time, given they would be racing to acquire that lock every time a message needs to be sent (?)

w.r.t wether or not a deadlock is possible, I think @mzumsande may be able to tell you better given he was the one pointing them out initially, but if we want to enforce a certain ordering of acquiring them I think my suggestion (or something on those lines) may be worth adding. I'm even happy to open a PR for that

Reply:
> Something along the lines of:
> 
> ```python
> def p2p_lock(self):
>     assert not self._send_lock.locked()
>     return p2p_lock
> ```

I think that this is not right (I tried it out, and it asserts everywhere :sweat_smile:):
To prevent deadlocks, the lock order must be preserved only within each thread:
- It's possible that a given thread first takes `p2p_lock`, and then `send_lock`. See for example: `on_message` -> `p2p_lock` taken -> call `on_inv` -> call `send_message` -> `_send_lock` taken
- What must be disallowed is that any thread can do the reverse: First take `_send_lock`, and then take `p2p_lock`, because it would lead to a deadlockif one thread did that, and another thread did the reverse. 
- I don't think that this is possible because I looked at the one spot where `send_lock` is taken (it calls `self.build_message` and `self.send_raw_message`, neither of which take `p2p_lock`) 
- But it is perfectly allowed that Thread 1 take `p2p_lock`, and immediately after Thread 2 takes `_send_lock` (which would trigger the assert).

Reply:
You're right. I don't think there is any easy way of keeping track of who's thread is holding what lock in Python though, so I don't think there a straightforward way of addressing this :(